### DISCUSSION TITLE
An error in RepeatableSampler. Series -> numpy

### DISCUSSION THREAD

**Discussion by Jeriousman** [RUCAIBox/RecBole#1214] [Q&A] [answered]
I was digging your sampler code. How I did was as this. First I called create_samplers().
```
from recbole.data import create_dataset, data_preparation, save_split_dataloaders, load_split_dataloaders, create_samplers
train_sampler, valid_sampler, test_sampler = create_samplers(config, dataset, data_preparation(config, dataset))
```
Then with valid_sampler, I was checking the samplercode. As my valid_sampler is <__main__.RepeatableSampler at 0x7f16253faa10>,  I was checking RepeatableSampler.
Here in     ```def _get_candidates_list(self)```, I get this error below.
```AttributeError: 'Series' object has no attribute 'numpy'```
So I replaced .numpy() -> .to_numpy() and it works fine. My autopsy process for the code is wrong? Or is it something that should be fixed?

Thank you in advance.

**Comment by Wicknight** [comment, marked as answer]
@Jeriousman Hello, thanks for your attention to RecBole!
For this problem, I think there is something wrong with your code.

First, you didn't give the code for how the dataset was created. Judging from the error message, the problem seems to be that the attribute of '**inter_feat**' in dataset is **not converted from DataFrame to Interaction**. I think this is because you didn't call the function of 'dataset.build()' after creating the dataset.

Second, I don't understand why you use ‘data_preparation(config, dataset)’ as the third parameter of create_samplers() function.

**Reply by Jeriousman** [reply]
Sorry for not being so clear. I have done changing format. So basically I used _dataframe_to_interaction function. and your second question, because creat_samplers needs built_datasets which is described as ```built_datasets (list of Dataset): A list of split Dataset, which contains dataset for training, validation and testing.```. If I am wrong about the built_dataset being the outputs of data_preparation(config, dataset), can you tell me which are right built_datasets in detail?

Below is just create_samplers function code for your convenience.


```
def create_samplers(config, dataset, built_datasets):
    """Create sampler for training, validation and testing.

    Args:
        config (Config): An instance object of Config, used to record parameter information.
        dataset (Dataset): An instance object of Dataset, which contains all interaction records.
        built_datasets (list of Dataset): A list of split Dataset, which contains dataset for
            training, validation and testing.

    Returns:
        tuple:
            - train_sampler (AbstractSampler): The sampler for training.
            - valid_sampler (AbstractSampler): The sampler for validation.
            - test_sampler (AbstractSampler): The sampler for testing.
    """
    phases = ['train', 'valid', 'test']
    train_neg_sample_args = config['train_neg_sample_args']
    eval_neg_sample_args = config['eval_neg_sample_args']
    sampler = None
    train_sampler, valid_sampler, test_sampler = None, None, None

    if train_neg_sample_args['strategy'] != 'none':
        if not config['repeatable']:
            sampler = Sampler(phases, built_datasets, train_neg_sample_args['distribution'])
        else:
            sampler = RepeatableSampler(phases, dataset, train_neg_sample_args['distribution'])
        train_sampler = sampler.set_phase('train')

    if eval_neg_sample_args['strategy'] != 'none':
        if sampler is None:
            if not config['repeatable']:
                sampler = Sampler(phases, built_datasets, eval_neg_sample_args['distribution'])
            else:
                sampler = RepeatableSampler(phases, dataset, eval_neg_sample_args['distribution'])
        else:
            sampler.set_distribution(eval_neg_sample_args['distribution'])
        valid_sampler = sampler.set_phase('valid')
        test_sampler = sampler.set_phase('test')

    return train_sampler, valid_sampler, test_sampler
```

**Reply by Wicknight** [reply]
@Jeriousman 
First, if you call the ‘_dataframe_to_interaction’ function correctly, the error message should not be reported. Because this error is obviously caused by the data type. Normally, the 'self.dataset.inter_feat[self.iid_field]' in the function here should have been changed to tensor type. Therefore, I recommend that you debug to see if the feats in the dataset have been converted to tensor types correctly.
Second, the outputs of data_preparation(config, dataset) are a list of split dataloaders, while creat_samplers needs a list of split datasets. The two are different. You can just use 'dataset.build()' to get built datasets.

**Reply by Jeriousman** [reply]
Thank you for your feedback! Much appreciated.
